osu.Game.Rulesets.Osu/Mods/OsuModBubbles.cs

```diff
@@ -0,0 +1,237 @@
+                        FadeTime = bubbleFade,
+                        Colour = drawableOsuHitObject.AccentColour.Value,
+                        IsHit = drawableOsuHitObject.IsHit,
+                    };
+                    bubbleContainer.Add(bubble);
+                }
+
+                Vector2 getPosition()
+                {
+                    switch (drawableOsuHitObject)
+                    {
+                        // SliderHeads are derived from HitCircles,
+                        // so we must handle them before to avoid them using the wrong positioning logic
+                        case DrawableSliderHead:
+                            return drawableOsuHitObject.HitObject.Position;
+
+                        // Using hitobject position will cause issues with HitCircle placement due to stack leniency.
+                        case DrawableHitCircle:
+                            return drawableOsuHitObject.Position;
+
+                        default:
+                            return drawableOsuHitObject.HitObject.Position;
+                    }
+                }
+            };
+
+            drawableObject.OnRevertResult += (drawable, _) =>
+            {
+                if (drawable.HitObject is SpinnerTick or Slider) return;
+
+                BubbleDrawable? lastBubble = bubbleContainer.OfType<BubbleDrawable>().LastOrDefault();
+
+                lastBubble?.ClearTransforms();
+                lastBubble?.Expire();
+            };
+        }
+
+        // Makes the slider border coloured on all skins (for aesthetics)
+        private void applySliderState(DrawableSlider slider) =>
+            ((PlaySliderBody)slider.Body.Drawable).BorderColour = slider.AccentColour.Value;
+
+        #region Pooled Bubble drawable
+
+        private partial class BubbleDrawable : PoolableDrawable
+        {
+            private readonly Box colourBox;
+            private readonly CircularContainer content;
+
+            public BubbleInfo Info { get; set; }
+
+            public BubbleDrawable()
+            {
+                Origin = Anchor.Centre;
+                InternalChild = content = new CircularContainer
+                {
+                    Anchor = Anchor.Centre,
+                    Origin = Anchor.Centre,
+                    RelativeSizeAxes = Axes.Both,
+                    MaskingSmoothness = 2,
+                    BorderThickness = 0,
+                    BorderColour = Colour4.White,
+                    Masking = true,
+                    EdgeEffect = new EdgeEffectParameters
+                    {
+                        Type = EdgeEffectType.Shadow,
+                        Radius = 3,
+                        Colour = Colour4.Black.Opacity(0.05f),
+                    },
+                    Child = colourBox = new Box { RelativeSizeAxes = Axes.Both, }
+                };
+            }
+
+            protected override void PrepareForUse()
+            {
+                Alpha = 1;
```

Comment:
Should not be required as you're using `FadeOutFromOne`.

Reply:
it is necessary. since its not used at the start of the transforms

Reply:
Generally I think it reads better to set these initial values as part of the transform sequence, ie.

```csharp
this.FadeTo(1).FadeOut(500);
```

Reply:
I'd push back softly in this case since I feel there's a separation to be made between stuff that is  "resetting" the `Drawable` for pool reuse, and then transforms being used to create the animation. At least that makes sense to me in my head, but yeah not strongly opposed to this change.